File: src/types/email.ts
/**
 * Email types for the email client
 */

import type { AccountColor } from './account'

export interface EmailAddress {
  name: string
  email: string
}

export interface Attachment {
  id: string
  name: string
  size: number
  mimeType: string
  url?: string
}

export interface Label {
  id: string
  name: string
  color: string
}

export type SystemFolderType = 'inbox' | 'drafts' | 'sent' | 'spam' | 'trash' | 'archive' | 'starred'
export type FolderType = SystemFolderType | string // string for custom folders

export interface CustomFolder {
  id: string
  name: string
  color: string
  parentId: string | null
  order: number
  createdAt: Date
  updatedAt: Date
}

export interface Email {
  id: string
  accountId: string
  accountColor: AccountColor
  from: EmailAddress
  to: EmailAddress[]
  cc?: EmailAddress[]
  bcc?: EmailAddress[]
  subject: string
  preview: string
  body: string
  date: Date
  isRead: boolean
  isStarred: boolean
  hasAttachment: boolean
  attachments: Attachment[]
  folder: FolderType
  labels: string[]
  threadId: string
  replyToId?: string
  forwardedFromId?: string
  isDraft: boolean
  scheduledSendAt?: Date
}

export interface EmailCreateInput {
  accountId: string
  to: EmailAddress[]
  cc?: EmailAddress[]
  bcc?: EmailAddress[]
  subject: string
  body: string
  attachments?: Attachment[]
  replyToId?: string
  forwardedFromId?: string
  scheduledSendAt?: Date
}

export interface EmailUpdateInput {
  isRead?: boolean
  isStarred?: boolean
  folder?: FolderType
  labels?: string[]
}

export interface EmailSearchQuery {
  text?: string
  from?: string
  to?: string
  subject?: string
  hasAttachment?: boolean
  isStarred?: boolean
  isRead?: boolean
  folder?: FolderType
  labels?: string[]
  dateFrom?: Date
  dateTo?: Date
  accountId?: string
}

/**
 * Input type for composing and sending emails
 * Used by the compose window and sendEmail action
 */
export interface ComposeEmailInput {
  /** Existing email ID if editing a draft */
  id?: string
  /** Account to send from */
  accountId: string
  /** Sender information (typically set from account) */
  from?: EmailAddress
  /** Recipients */
  to: EmailAddress[]
  /** CC recipients */
  cc?: EmailAddress[]
  /** BCC recipients */
  bcc?: EmailAddress[]
  /** Email subject */
  subject: string
  /** Email body (HTML) */
  body: string
  /** Attachments */
  attachments?: Attachment[]
  /** Thread context */
  threadId?: string
  /** Original email ID if replying */
  replyToId?: string
  /** Original email ID if forwarding */
  forwardedFromId?: string
  /** True if this is an existing draft being sent */
  isDraft?: boolean
  /** Account color for display */
  accountColor?: AccountColor
}

// System folder definitions
export const SYSTEM_FOLDERS: SystemFolderType[] = [
  'inbox',
  'drafts',
  'sent',
  'spam',
  'trash',
  'archive',
  'starred',
]

export function isSystemFolder(folder: string): folder is SystemFolderType {
  return SYSTEM_FOLDERS.includes(folder as SystemFolderType)
}

File: backend/penguin_mail/api/schemas/email.py
from datetime import datetime
from typing import Literal

from ninja import Field, Schema
from pydantic import model_serializer

# Valid folder names — mirrors FolderType model choices.
ValidFolder = Literal[
    "inbox",
    "drafts",
    "sent",
    "spam",
    "trash",
    "archive",
    "starred",
    "snoozed",
    "scheduled",
]


class EmailAddressOut(Schema):
    name: str
    email: str


class AttachmentBrief(Schema):
    id: str
    name: str
    size: int
    mimeType: str
    url: str | None = None


class EmailOut(Schema):
    id: str
    accountId: str
    accountColor: str
    from_: EmailAddressOut = Field(serialization_alias="from")
    to: list[EmailAddressOut]
    cc: list[EmailAddressOut] = []
    bcc: list[EmailAddressOut] = []
    subject: str
    preview: str
    body: str
    date: datetime
    isRead: bool
    isStarred: bool
    hasAttachment: bool
    attachments: list[AttachmentBrief] = []
    folder: str
    labels: list[str] = []
    threadId: str | None = None
    replyToId: str | None = None
    forwardedFromId: str | None = None
    isDraft: bool
    scheduledSendAt: datetime | None = None
    snoozeUntil: datetime | None = None
    snoozedFromFolder: str | None = None

    class Config:
        json_schema_extra = {"properties": {"from": {"$ref": "#/$defs/EmailAddressOut"}}}

    @model_serializer(mode="wrap")
    def _serialize(self, handler):
        d = handler(self)
        if "from_" in d:
            d["from"] = d.pop("from_")
        return d

    @staticmethod
    def from_model(email) -> "EmailOut":
        recipients = email.recipients.all()
        to_list = [EmailAddressOut(name=r.name, email=r.address) for r in recipients if r.kind == "TO"]
        cc_list = [EmailAddressOut(name=r.name, email=r.address) for r in recipients if r.kind == "CC"]
        bcc_list = [EmailAddressOut(name=r.name, email=r.address) for r in recipients if r.kind == "BCC"]

        attachments = [
            AttachmentBrief(
                id=str(a.uuid),
                name=a.name,
                size=a.size,
                mimeType=a.mime_type,
                url=a.file.url if a.file else None,
            )
            for a in email.attachments.all()
        ]

        label_ids = [str(l.uuid) for l in email.labels.all()]

        return EmailOut(
            id=str(email.uuid),
            accountId=str(email.account.uuid),
            accountColor=email.account.color,
            from_=EmailAddressOut(name=email.sender_name, email=email.sender_email),
            to=to_list,
            cc=cc_list,
            bcc=bcc_list,
            subject=email.subject,
            preview=email.preview,
            body=email.body,
            date=email.created_at,
            isRead=email.is_read,
            isStarred=email.is_starred,
            hasAttachment=email.has_attachment,
            attachments=attachments,
            folder=email.folder,
            labels=label_ids,
            threadId=str(email.thread_id) if email.thread_id else None,
            replyToId=str(email.reply_to.uuid) if email.reply_to else None,
            forwardedFromId=str(email.forwarded_from.uuid) if email.forwarded_from else None,
            isDraft=email.is_draft,
            scheduledSendAt=email.scheduled_send_at,
            snoozeUntil=email.snooze_until,
            snoozedFromFolder=email.snoozed_from_folder,
        )


class EmailAddressIn(Schema):
    name: str = ""
    email: str


class EmailCreateIn(Schema):
    accountId: str
    to: list[EmailAddressIn]
    cc: list[EmailAddressIn] = []
    bcc: list[EmailAddressIn] = []
    subject: str = ""
    body: str = ""
    replyToId: str | None = None
    forwardedFromId: str | None = None
    scheduledSendAt: datetime | None = None


class EmailUpdateIn(Schema):
    isRead: bool | None = None
    isStarred: bool | None = None
    folder: ValidFolder | None = None
    labels: list[str] | None = None


BulkOperation = Literal[
    "markRead",
    "markUnread",
    "star",
    "unstar",
    "archive",
    "delete",
    "deletePermanent",
    "move",
    "addLabel",
    "removeLabel",
]


class BulkOpIn(Schema):
    ids: list[str]
    operation: BulkOperation
    folder: ValidFolder | None = None
    labelIds: list[str] | None = None


class LabelOpIn(Schema):
    labelIds: list[str]
